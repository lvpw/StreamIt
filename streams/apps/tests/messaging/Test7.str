void->void pipeline Test7 {
    portal<Counter> p1, p2;
    add Source(p1);
    add PassThrough(p2);
    add Counter to p1, p2;
    add Output;
}

void->int filter Source(portal<Counter> p) {
    int i=100;
    work push 1 {
	i++;
	push(i);
	if (i==110) {
	    i = 100;
	    p.incVal() [0:0];
	}
    }
}

int->int filter PassThrough(portal<Counter> p) {
    int i=100;
    work push 1 pop 1 {
        i++;
        push(pop());
	if (i==110) {
            i = 100;
            p.incVal() [0:0];
        }
    }
}

int->int filter Counter {
    int val = 0;
    
    work push 2 pop 1 {
	push(pop());
	push(val);
    }

    handler incVal() {
	val++;
    }
}

int->void filter Output {
  work pop 1 {
    println(pop());
  }
}

/* OUTPUT
cagfarm-44 % strc -library Test7.str | & more
101
0
102
0
103
0
104
0
105
0
106
0
107
0
108
0
109
0
Message delivery
Message delivery
110
2
101
2
102
2
103
2
104
2
105
2
106
2
107
2
108
2
109
2
Message delivery
Message delivery
110
4
101
4
102
4
103
4
104
4
105
4
106
4
107
4
108
4
109
4
Message delivery
Message delivery
110
6
*/
