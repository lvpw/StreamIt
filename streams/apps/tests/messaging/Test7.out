101
0
102
0
103
0
104
0
105
0
106
0
107
0
108
0
109
0
110
2
101
2
102
2
103
2
104
2
105
2
106
2
107
2
108
2
109
2
110
4
101
4
102
4
103
4
104
4
105
4
106
4
107
4
108
4
109
4
110
6
101
6
102
6
103
6
104
6
105
6
106
6
107
6
108
6
109
6
110
8
101
8
102
8
103
8
104
8
105
8
106
8
107
8
108
8
109
8
110
10
101
10
102
10
103
10
104
10
105
10
106
10
107
10
108
10
109
10
110
12
101
12
102
12
103
12
104
12
105
12
106
12
107
12
108
12
109
12
110
14
101
14
102
14
103
14
104
14
105
14
106
14
107
14
108
14
109
14
110
16
101
16
102
16
103
16
104
16
105
16
106
16
107
16
108
16
109
16
110
18
101
18
102
18
103
18
104
18
105
18
106
18
107
18
108
18
109
18
110
20
